File: Recursion and Backtracking/NQueen/nQueen.java
/***
 * Code Written By: Nishkarsh Makhija
 * Code Written for: Learn Algorithm
 * Insta Handle : learn_algorithm  
**/

class nQueen {

    static int N = 4;

    static boolean solve_for(int board[], int currentRow)
    {
        if(currentRow >= N)
            return true;

        for(int currentCol = 0; currentCol < N; currentCol++)
        {
            if(is_safe(board, currentRow, currentCol)){
                board[currentRow] = currentCol;
                if(solve_for(board, currentRow+1) )
                    return true;
            }
        }
        return false;
    }

    static boolean is_safe(int board[], int currentRow, int currentCol)
    {
        for(int row = 0; row <= currentRow; row++)
        {
            if(board[row] == currentCol || Math.abs(board[row] - currentCol) == Math.abs(row - currentRow) )
                return false;
        }
        return true;
    }

    static void displayBoard(int board[])
    {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++){
                if(board[i] == j){
                    System.out.print("Q ");
                }
                else{
                    System.out.print("_ ");
                }
            }
            System.out.println("");
        }
    }

    public static void main(String args[])
    {
        int board[] = {-1, -1, -1, -1};

        solve_for(board, 0);
        displayBoard(board);
    }

}






    

    

File: Recursion and Backtracking/NQueen/nQueen.py
'''
 * Code Written By: Nishkarsh Makhija
 * Code Written for: Learn Algorithm
 * Insta Handle : learn_algorithm  
'''

N = 8
def solve_for(board, currentRow=0):
    if currentRow >= N:
        return True
        
    for currentCol in range(N):
        if is_safe(board, currentRow, currentCol):
            board[currentRow] = currentCol
            if(solve_for(board, currentRow+1) ):
                return True
    return False
    
def is_safe(board, currentRow, currentCol):
    for row in range(currentRow):
        if(board[row] == currentCol or abs(board[row] - currentCol) == abs(row - currentRow)):
            return False
            
    return True

    
def displayBoard(board):
    for row in range(N):
        for col in range(N):
            if board[row] == col:
                print("Q",end=" ")
            else:
                print("_",end=" ")
        print()

def main():
    board = [-1 for _ in range(N)]
    solve_for(board)
    displayBoard(board)
main()


